### ISSUE TITLE
gnt-cluster verify reports ERROR:ENODEHOOKS when a node is offline

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#798]
Originally reported of Google Code with ID 742.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

# gnt-cluster --version; gnt-cluster version; hspace --version
gnt-cluster (ganeti v2.10.0) 2.10.0
Software version: 2.10.0
Internode protocol: 2100000
Configuration format: 2100000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.10.0
hspace (ganeti) version v2.10.0
compiled with ghc 6.12
running on linux x86_64

<b>What distribution are you using?</b>

Debian squeeze.

<b>What steps will reproduce the problem?</b>
1. gnt-node modify -O yes $node
2. Ensure node is actually unreachable (ssh $node service ganeti stop ganeti-noded is enough)
3. gnt-cluster verify --error-codes.

<b>What is the expected output? What do you see instead?</b>

Expected output: gnt-cluster verify passes with no errors related to the node being offline.
Actual output:
Mon Mar  3 11:45:28 2014 * Other Notes
Mon Mar  3 11:45:28 2014   - NOTICE: 1 offline node(s) found.
Mon Mar  3 11:45:32 2014  - WARNING: Communication failure to node $OFFLINENODE: Error 7: Failed connect to $OFFLINENODE:1811; Resource temporarily unavailable
Mon Mar  3 11:45:32 2014 * Hooks Results
Mon Mar  3 11:45:32 2014   - ERROR:ENODEHOOKS:node:$OFFLINENODE:Communication failure in hooks execution: Error 7: Failed connect to $OFFLINENODE:1811; Resource temporarily unavailable

<b>Please provide any additional information below.</b>

I tried to track down the problem in Ganeti to make sure it wasn't a problem with our local setup. The cause seems to be that _CheckConfigNode in lib/rpc.py does not do offline detection for nodes addressed by their hostname instead of UUID, and it appears that all job hooks address nodes by their hostnames (through the to_name lambda in HooksMaster.BuildFromLu in lib/hooksmaster.py).

I've attached a diff containing a dirty workaround for the problem, which fixes the immediate issue. I'm not sure if the proper fix should be to switch from names to UUIDs for job hooks or changing the config API to be able to look up by name again (or something else entirely) so I'll leave that to you :).
```
Originally added on 2014-03-03 11:59:50 +0000 UTC.

Attachments: 
[gnt-cluster_verify_hacky_workaround.patch](https://storage.googleapis.com/google-code-attachments/ganeti/issue-742/comment-0/gnt-cluster_verify_hacky_workaround.patch)

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-03-03 12:12:22 +0000 UTC.

Added to Milestone: Release2.10

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-03-03 13:33:49 +0000 UTC.

Added Labels: Priority-High
Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
The reason why _CheckConfigNode can handle either node UUID's or node names is LUNodeAdd. The newly added node therein has no UUID when the BuildHooksNodes() method is invoked, so the new node is addressed by its node name instead.

Having _CheckConfigNode fall back to the node name was probably not the wisest solution. At this point in the control flow it's pretty impossible (or really hacky) to figure out whether it's ok to talk to the node by name or not.

The patch which is attached will probably break node adds, but if you return the name instead of None in foo it would probably work. But in any case, it's not a clean solution.

IMHO, the best approach to fix this would be to accept node UUIDs only in _CheckConfigNode() and fix LUNodeAdd to only use node UUIDs in BuildHooksNodes(). The problem there is that BuildHooksNodes() is called before the LU is actually executed, so the node UUID is not known at this point.
I see two potential fixes for that:
 * Reserve an UUID before BuildHooksNodes() is called and use this UUID then during Exec()
 * Add a AppendPostHooksNodes() method which is called before POST hooks are called, so LUNodeAdd could add the UUID of the newly added node during this method.



```
Originally added on 2014-03-03 14:24:51 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-03-03 14:26:55 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed with:

commit ff1c051bb3fb0bf6ca10cdeabb3221c10970da68
Author: Thomas Thrainer <thomasth@google.com>
Date:   Mon Mar 3 15:57:12 2014 +0100

    Use node UUIDs for executing LU hooks
    
    LUNodeAdd, the only LU using a node name still, is changed to overwrite
    PreparePostHookNodes() and use node UUIDs only as well.
    This allows to remove the support for 3-tuples as results of
    BuildHooksNodes() and removes the translation to node names.
    
    Fixes issue 742.
    
    Signed-off-by: Thomas Thrainer <thomasth@google.com>
    Reviewed-by: Michele Tartara <mtartara@google.com>

commit 237a833cd1d33b9454420d8a55234aa2dff135ce
Author: Thomas Thrainer <thomasth@google.com>
Date:   Mon Mar 3 15:49:57 2014 +0100

    Add PreparePostHookNodes to LUs
    
    This method can be used to alter the list of node UUIDs on which post
    hooks are executed. PreparePostHookNodes is called after Exec, so LUs
    can use data only known after the execution of the LU.
    
    Signed-off-by: Thomas Thrainer <thomasth@google.com>
    Reviewed-by: Michele Tartara <mtartara@google.com>
```
Originally added on 2014-03-04 09:42:03 +0000 UTC.

Mentioned Issue with GoogleCode ID 742 has been migrated to GitHub with ID #798.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released in 2.10.1
```
Originally added on 2014-03-05 13:55:53 +0000 UTC.

Changed State: Released
